File: 3.x.x/crates-utopia/messaging/src/adapter/sms/fast2sms.rs
//! PHP `Utopia\Messaging\Adapter\SMS\Fast2SMS`.

use serde_json::json;

use super::geosms::CallingCode;
use super::TYPE;
use crate::adapter::{expect_sms, Adapter, AdapterBase, SendResult};
use crate::error::MessagingError;
use crate::message::{Message, MessageKind};
use crate::messages::SMS;
use crate::response::{Response, ResponseData};

/// PHP `Adapter\SMS\Fast2SMS`.
#[derive(Debug)]
pub struct Fast2SMS {
    base: AdapterBase,
    api_key: String,
    sender_id: String,
    message_id: String,
    use_dlt: bool,
}

impl Fast2SMS {
    /// PHP `__construct($apiKey, $senderId = '', $messageId = '', $useDLT = false)`.
    #[must_use]
    pub fn new(
        api_key: impl Into<String>,
        sender_id: impl Into<String>,
        message_id: impl Into<String>,
        use_dlt: bool,
    ) -> Self {
        Self {
            base: AdapterBase::default(),
            api_key: api_key.into(),
            sender_id: sender_id.into(),
            message_id: message_id.into(),
            use_dlt,
        }
    }

    fn remove_country_code(number: &str) -> String {
        let digits: String = number.chars().filter(char::is_ascii_digit).collect();
        if let Some(code) = CallingCode::from_phone_number(number) {
            return digits.get(code.len()..).unwrap_or("").to_string();
        }
        digits
    }

    fn process_sms(&self, message: &SMS) -> ResponseData {
        let numbers: Vec<String> = message
            .get_to()
            .iter()
            .map(|n| Self::remove_country_code(n))
            .collect();
        let mut payload = json!({
            "numbers": numbers.join(","),
            "flash": 0,
        });
        if self.use_dlt {
            payload["route"] = json!("dlt");
            payload["sender_id"] = json!(self.sender_id);
            payload["message"] = json!(self.message_id);
            payload["variables_values"] = json!(message.get_content());
        } else {
            payload["route"] = json!("q");
            payload["message"] = json!(message.get_content());
        }
        let result = self.request_default(
            "POST",
            "https://www.fast2sms.com/dev/bulkV2",
            &[
                format!("authorization: {}", self.api_key),
                "Content-Type: application/json".into(),
                "Accept: application/json".into(),
            ],
            Some(payload),
        );
        let mut response = Response::new(TYPE);
        let ret = result
            .response
            .get("return")
            .and_then(serde_json::Value::as_bool);
        if result.status_code == 200 && ret == Some(true) {
            response.set_delivered_to(message.get_to().len() as i64);
            for to in message.get_to() {
                response.add_result(to, "");
            }
        } else {
            // PHP: `$res['message'] ?? 'Unknown error' . ' Status Code: ' . ...`
            // `.` binds tighter than `??`.
            let error = if let Some(msg) = result.response.get("message").and_then(|v| v.as_str()) {
                msg.to_string()
            } else {
                let status = result
                    .response
                    .get("status_code")
                    .map_or_else(|| "Unknown".into(), ToString::to_string);
                format!("Unknown error Status Code: {status}")
            };
            for to in message.get_to() {
                response.add_result(to, error.clone());
            }
        }
        response.to_array()
    }
}

impl Adapter for Fast2SMS {
    fn get_name(&self) -> &'static str {
        "Fast2SMS"
    }
    fn get_type(&self) -> &'static str {
        TYPE
    }
    fn get_message_type(&self) -> MessageKind {
        MessageKind::SMS
    }
    fn get_max_messages_per_request(&self) -> usize {
        1000
    }
    fn base(&self) -> &AdapterBase {
        &self.base
    }
    fn process(&self, message: &dyn Message) -> Result<SendResult, MessagingError> {
        Ok(SendResult::Response(self.process_sms(expect_sms(message)?)))
    }
}

File: packages/messaging/src/Adapter/SMS/Fast2SMS.php
<?php

namespace Utopia\Messaging\Adapter\SMS;

use Utopia\Messaging\Adapter\SMS as SMSAdapter;
use Utopia\Messaging\Adapter\SMS\GEOSMS\CallingCode;
use Utopia\Messaging\Messages\SMS as SMSMessage;
use Utopia\Messaging\Response;

/**
 * Fast2SMS Adapter for sending SMS messages using Fast2SMS API
 *
 * This adapter supports both DLT (Distributed Ledger Technology) and Quick SMS routes
 * For API documentation, see: https://docs.fast2sms.com
 */
class Fast2SMS extends SMSAdapter
{
    protected const NAME = 'Fast2SMS';
    protected const API_ENDPOINT = 'https://www.fast2sms.com/dev/bulkV2';

    /**
     * Create a new Fast2SMS adapter instance
     *
     * @param string $apiKey Your Fast2SMS API authorization key
     * @param string $senderId Your 3-6 letter DLT approved Sender ID (e.g., "FSTSMS"). Required for DLT route
     * @param string $messageId Your DLT approved Message ID template. Required for DLT route
     * @param bool $useDLT Whether to use DLT route (true) or Quick SMS route (false)
     *
     * @see https://docs.fast2sms.com/#dlt-sms
     * @see https://docs.fast2sms.com/#quick-sms
     */
    public function __construct(
        private readonly string $apiKey,
        private readonly string $senderId = '',
        private readonly string $messageId = '',
        private readonly bool $useDLT = false,
    ) {
        parent::__construct();
    }

    /**
     * Get the name of the SMS adapter
     */
    public function getName(): string
    {
        return static::NAME;
    }

    /**
     * Get maximum number of messages that can be sent in a single request
     */
    public function getMaxMessagesPerRequest(): int
    {
        return 1000;
    }

    /**
     * Process the SMS message and send it using Fast2SMS API
     *
     * @param SMSMessage $message The SMS message to be processed
     * @return array<string, mixed> The response from the API
     */
    protected function process(SMSMessage $message): array
    {
        $numbers = array_map(
            $this->removeCountryCode(...),
            $message->getTo(),
        );
        $numbers = implode(',', $numbers);

        $payload = [
            'numbers' => $numbers,
            'flash' => 0,
        ];

        if ($this->useDLT) {
            $payload['route'] = 'dlt';
            $payload['sender_id'] = $this->senderId;
            $payload['message'] = $this->messageId;
            $payload['variables_values'] = $message->getContent();
        } else {
            $payload['route'] = 'q';
            $payload['message'] = $message->getContent();
        }

        $response = new Response($this->getType());
        $result = $this->request(
            method: 'POST',
            url: self::API_ENDPOINT,
            headers: [
                'authorization: ' . $this->apiKey,
                'Content-Type: application/json',
                'Accept: application/json',
            ],
            body: $payload,
        );

        $res = $result['response'];
        if ($result['statusCode'] === 200 && isset($res['return']) && $res['return'] === true) {
            $response->setDeliveredTo(\count($message->getTo()));
            foreach ($message->getTo() as $to) {
                $response->addResult($to);
            }
        } else {
            $errorMessage = $res['message'] ?? 'Unknown error' . ' Status Code: ' . ($res['status_code'] ?? 'Unknown');
            foreach ($message->getTo() as $to) {
                $response->addResult($to, $errorMessage);
            }
        }

        return $response->toArray();
    }

    /**
     * Removes country code from a phone number
     * Fast2SMS expects Indian phone numbers without the country code
     *
     * @param string $number Phone number with potential country code
     * @return string Phone number without country code
     */
    private function removeCountryCode(string $number): string
    {
        // Remove any non-digit characters
        $digits = preg_replace('/[^0-9]/', '', $number);

        $code = CallingCode::fromPhoneNumber($number);
        if ($code !== null) {
            return substr((string) $digits, \strlen($code));
        }

        return $digits;
    }
}
